src/engine/SpatialJoin.cpp

```diff
@@ -300,34 +376,150 @@ Result SpatialJoin::baselineAlgorithm() {
+
+  // Size of output table
   size_t numColumns = getResultWidth();
+  return std::make_tuple(resLeft, resRight, leftJoinCol, rightJoinCol,
+                         numColumns);
+}
+
+// ____________________________________________________________________________
+Result SpatialJoin::baselineAlgorithm() {
+  const auto [resLeft, resRight, leftJoinCol, rightJoinCol, numColumns] =
+      prepareJoin();
   IdTable result{numColumns, _executionContext->getAllocator()};
 
-  // cartesian product with a distance of at most maxDist between the two
-  // objects
+  // cartesian product between the two tables, pairs are restricted according to
+  // `maxDistance_` and `maxResults_`
   for (size_t rowLeft = 0; rowLeft < resLeft->size(); rowLeft++) {
+    // This priority queue stores the intermediate best results if `maxResults_`
+    // is used. Each intermediate result is stored as a pair of its `rowRight`
+    // and distance. Since the queue will hold at most `maxResults_ + 1` items,
+    // it is not a memory concern.
+    auto compare = [](std::pair<size_t, int64_t> a,
+                      std::pair<size_t, int64_t> b) {
+      return a.second > b.second;
+    };
+    std::priority_queue<std::pair<size_t, int64_t>,
+                        std::vector<std::pair<size_t, int64_t>>,
+                        decltype(compare)>
+        intermediate(compare);
+
+    // Inner loop of cartesian product
     for (size_t rowRight = 0; rowRight < resRight->size(); rowRight++) {
       Id dist = computeDist(resLeft, resRight, rowLeft, rowRight, leftJoinCol,
                             rightJoinCol);
-      if (dist.getDatatype() == Datatype::Int && dist.getInt() <= maxDist_) {
+
+      // Ensure `maxDist_` constraint
+      if (dist.getDatatype() != Datatype::Int || maxDist_ < 0 ||
+          dist.getInt() > maxDist_) {
+        continue;
+      }
+
+      // If there is no `maxResults_` we can add the result row immediately
+      if (maxResults_ < 0) {
         addResultTableEntry(&result, resLeft, resRight, rowLeft, rowRight,
                             dist);
+        continue;
+      }
+
+      // Ensure `maxResults_` constraint using priority queue
+      intermediate.push(std::pair{rowRight, dist.getInt()});
+      // Too many results? Drop the worst one
+      if (intermediate.size() > static_cast<size_t>(maxResults_)) {
+        intermediate.pop();
+      }
+    }
+
+    // If we are using the priority queue, we didn't add the results in the
+    // inner loop, so we do it now.
+    if (maxResults_ >= 0) {
+      size_t numResults = intermediate.size();
+      for (size_t item = 0; item < numResults; item++) {
+        // Get and remove largest item from priority queue
+        auto [rowRight, dist] = intermediate.top();
+        intermediate.pop();
+
+        addResultTableEntry(&result, resLeft, resRight, rowLeft, rowRight,
+                            Id::makeFromInt(dist));
```

Comment:
Ah, I am beginning to understand...
You also now have a baseline algorithm for "nearest neighbor".... which can be used for testing etc... nice, but we still should not represent "no max distance"as -1.